pkg/controller/installation/core_controller.go

```diff
@@ -1356,6 +1349,10 @@ func (r *ReconcileInstallation) Reconcile(ctx context.Context, request reconcile
 		instance.Status.ImageSet = imageSet.Name
 	}
 	instance.Status.Computed = &instance.Spec
+
+	// Update the installation status conditions from the tigerastatus.
+	r.updateInstallationStatus(instance, r.status.Conditions())
```

Comment:
Unless I'm missing something, I think there are one or two issues with this approach as it's currently formulated. Namely:

- this line is only hit after `r.Status.IsAvailable()` is true. So, we'll only update the status for explicit Degraded calls, and once Available, but never while progressing (which is the time between those states).
- The StatusManager updates the TigeraStatus object asynchronously from this loop - It's possible that the status manager will mark the TigeraStatus as degraded at some point in the future, and this controller won't recognize that.

The good thing about the way you have done this, though, is that it doesn't require any additional queries on the Installation API, and is all synchronous with the main controller loop - meaning we won't have conflicts between two controllers attempting to update the same object. 

We might want to sacrifice that in order to make sure the status is updated consistently with and at the same time as the TigeraStatus API.

I'm not sure yet the best way to do that. One option might be to adjust the StatusManager to take a new argument - a function which updates conditions. It can call that function as part of its syncloop.

e.g.,


```go
// Define a new type to use as the argument.
type StatusUpdater func([]TigeraStatusCondition) error
```

```go
// Add the new argument to the StatusManager.
func New(client client.Client, component string, kubernetesVersion *common.VersionInfo, updateFunc StatusUpdater) StatusManager {
```

And then you can replace this line: https://github.com/tigera/operator/blob/6262d7798f5ca2e6a09e85b262eaa1cb91f59fdc/pkg/controller/status/status.go#L700

With `err := sm.updateFunc(ts.Status.Conditions)`

In order for this to work, we'd also need to move the rest of the status updating logic into that function as well - otherwise we'd have two goroutines trying to update the status at the same time, which would result in conflict errors being returned from the API server. It also has the downside of not having access to all of the fields and logic that may have occurred in the Reconcile() loop (e.g., imageSet).

I suppose another alternative would be to have the controller watch the TigeraStatus resource, so that any time the TigeraStatus resource changes we can query it, and update the Installation CR status based on what we find. 

Before you go down this path would want to hear from @tmjd to see if he has any other ideas or thoughts on these approaches.

Reply:
"The StatusManager updates the TigeraStatus object asynchronously from this loop - It's possible that the status manager will mark the TigeraStatus as degraded at some point in the future, and this controller won't recognize that."

I was under the impression that the status manager does not set a `Degraded` condition by itself and that the controller is the authority on this condition.
At the same time, the status manager is the authority on the `Progressing` condition. 
Lastly, `Available` is the result of `!(Degraded | Progressing)`. 
Vara's approach is mostly correct, but once the status moves to `Progressing=True`, a lagging effect appears in the Installation.Status, because this is not noticed until the next reconciliation. At that time `Progressing` may be false again and therefore, `Progressing=true` may not be noticed at all.
I am writing this because I am personally wondering how big of a problem it is when missing `Progressing=true` is if it is short-lived. 

Rather than having two controllers update the Installation.Status, could the status manager send a message to a channel that the core_controller configures the status manager with to update the status?

Reply:
I think the TigeraStatus watching strategy seems like a possibly nice solution as far as keeping resources only updated from one routine (and not needing to fetch the resource before updating). We could possibly even have the Reconcile check that it is a tigerastatus update and skip the rest of the reconcile and only update the resource status.

I'm not understanding how the StatusUpdater could work, since as you mentioned it would need other field info from the Reconcile.

What about if we had a channel where the controller (Reconcile) would send the resource to the StatusManager and the StatusManager was responsible for making the call to update the Status?

@rene-dekker the Progressing portion, especially for calico, may be a significant time on a sizeable cluster.
I'm kind of suggesting something similar to your suggestion but the other way.
I'm not sure I understand what you're fully suggesting because I don't know what would be able to listen on the channel for changes from the Status manager, it would still be at the mercy of the Reconcile being triggered.

Reply:
> I was under the impression that the status manager does not set a Degraded condition by itself

@rene-dekker the status manager can mark a CR as degraded independently of the controller, if for example there are pods that are failing in some way: https://github.com/tigera/operator/blob/6262d7798f5ca2e6a09e85b262eaa1cb91f59fdc/pkg/controller/status/status.go#L467-L468

> I'm not understanding how the StatusUpdater could work, since as you mentioned it would need other field info from the Reconcile.

It would be awkward for some fields - we'd need to cache them on the Installation and then use a callback to a bound function on the reconciler that would set the fields appropriately. Probably not my favorite approach.

> What about if we had a channel where the controller (Reconcile) would send the resource to the StatusManager and the StatusManager was responsible for making the call to update the Status?

The problem is that the controller / reconciler doesn't know all the times that it should send the resource to the StatusManager - the StatusManager independently updates conditions even if no reconcile has occurred, and so the trigger needs to come from the other direction. 

> I think the TigeraStatus watching strategy seems like a possibly nice solution as far as keeping resources only updated from one routine (and not needing to fetch the resource before updating). We could possibly even have the Reconcile check that it is a tigerastatus update and skip the rest of the reconcile and only update the resource status.

Yep exactly - that's why I'm leaning towards this approach right now. Seems to have a solid separation of concerns.

Reply:
Sounds good. So to sum up there are some take-aways:
- You can add a watch for TigeraStatus with name "calico" in the func `add(...)`  core_controllers. This will make the controller reconcile whenever the TigeraStatus is modified in the datastore.
- Then in the start of the `Reconcile(...)` func, I think you can use the request parameter to figure out if the TigeraStatus was updated. I'm not sure if there might be a better way, since` reconcile.Request` only has the namespace and name to find out what was modified.

Reply:
> > What about if we had a channel where the controller (Reconcile) would send the resource to the StatusManager and the StatusManager was responsible for making the call to update the Status?

> The problem is that the controller / reconciler doesn't know all the times that it should send the resource to the StatusManager - the StatusManager independently updates conditions even if no reconcile has occurred, and so the trigger needs to come from the other direction.

I was imagining the StatusManager would hold the last sent resource (Installation CR) and use the last sent one during any updates until it received a new one. We might need to send it multiple times during a Reconcile, one at the start and one at the finish. If we include `observedGeneration` then that would include the information the StatusManager would need to determine the proper state of the "Reconciling" condition.